### PR TITLE
Dynamically set `max_new_tokens` based on output feature length, GMSL and model window size

### PR DESCRIPTION
**Pull request by arnavgarg1**
TLDR; This fixes a bug in `model.evaluation()` that caused underreporting of performance metrics in a majority of cases. By setting max_new_tokens to the largest possible value _needed_, we ensure that the metric number from `model.evaluate()` are a good representation of true model performance.

---

This PR updates the config validation setter methods to ensure that the `generation.max_new_tokens` parameter in the model configuration is set correctly based on the maximum sequence length of the output features. This ensures accurate token generation for LLM model types. It also includes handling for different cases and provides informative logging for configuration changes. 

Specifically:
1. If the user sets max_new_tokens in generation to a value != the default value, it is respected as is.
2. Otherwise, we fall back to the max_sequence_length set for the `output` feature since that is the max number of tokens it learns during fine-tuning.
3. Otherwise, we fall back to the global_max_sequence_length if neither case 1 or case 2 are met. This is an overestimation of the number of required tokens, but it is acceptable for now.
4. Finally, if none of case 1, 2 or 3 are met, set max_new_tokens to the model's window size / 2. This feels like a good tradeoff of inference/evaluation time vs likely covering a majority of the total tokens allocated to the output feature. Can update this value in the future.

### CODE REVIEW

File: ludwig/schema/model_types/utils.py [hunk 1]
```diff
@@ -337,6 +346,66 @@ def set_llm_tokenizers(config: "ModelConfig") -> None:
...
+            max_possible_sequence_length, config.preprocessing.global_max_sequence_length
+        )
+
+    # It's possible that both max_sequence_length and global_max_sequence_length are not set, in which case
+    # we should fall back to the window size of the pretrained model. By this point, because of schema validation
+    # checks, we know that the base_model exists so we can safely grab the base model's config.
+    # TODO (Arnav): Figure out how to factor in rope scaling factor into this calculation.
+    if max_possible_sequence_length == default_max_sequence_length:
+        model_config = AutoConfig.from_pretrained(config.base_model)
+        # Determines the maximum length of the context (input + output tokens)
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply ludwig/schema/model_types/utils.py:390]: That's true, yep! Will create a shared helper instead!

File: ludwig/schema/model_types/utils.py [hunk 2]
```diff
@@ -337,6 +346,70 @@ def set_llm_tokenizers(config: "ModelConfig") -> None:
...
+        # If max_new_tokens is set, then we need to set max_length to None to ensure that the correct number of tokens
+        # are generated (input + output tokens), otherwise we will exceed the bounds of generation resulting in errors.
+        config.generation.max_new_tokens = None
+        config.generation.max_length = max_possible_sequence_length
+        logger.info(
+            f"Setting generation max_length to {max_possible_sequence_length} to correspond with the max sequence "
+            "length assigned to the output feature or the global max sequence length. This will ensure that the "
+            "correct number of tokens are generated at inference time. To override this behavior, set "
+            "`generation.max_length` to a different value in your Ludwig config."
+        )
```
Comment (arnavgarg1) [inline ludwig/schema/model_types/utils.py:402]: Will need to think through this part a bit more, I think the issue I was running into actually happens somewhere else

File: ludwig/schema/model_types/utils.py [hunk 3]
```diff
@@ -337,6 +346,61 @@ def set_llm_tokenizers(config: "ModelConfig") -> None:
             output_feature.decoder.fallback_label = output_feature.preprocessing.fallback_label
 
 
+def _set_generation_max_new_tokens(config: "ModelConfig") -> None:
+    """Sets the max_new_tokens parameter in the generation config to the max sequence length of the output
+    features.
+
+    This ensures that the generation config is set to the correct value for the LLM model type.
+    """
+    from transformers import AutoConfig
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply ludwig/schema/model_types/utils.py:355]: Agreed, moved these to the top - I forgot that transformers was always installed with even base Ludwig

File: tests/ludwig/utils/test_llm_utils.py [hunk 4]
```diff
@@ -57,6 +59,29 @@ def test_set_pad_token_already_exists():
     assert tokenizer.pad_token_id == 1
 
 
+class TestSetContextLen:
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply tests/ludwig/utils/test_llm_utils.py:62]: The idea in general is that since they're all testing the same function but different scenarios, it makes sense to either put them all in the same dedicated module for clarity or in some sort of container like a class.

[the remark this answers is not in the document]
Comment (justinxzhao) [reply tests/ludwig/utils/test_llm_utils.py:62]: I can see this being a useful way to organize tests particularly for very large test files. It's a bit more to maintain, but seems sufficiently lightweight.

**Review by justinxzhao — APPROVED** [review]

File: tests/ludwig/utils/test_llm_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply tests/ludwig/utils/test_llm_utils.py:62]: Got it, let me split them up!
